Added support for encoding and decoding tEXt/zTXt/iTXt chunks

sayantangkhan opened the pull request:
Added support for text chunks (as mentioned in [this issue](https://github.com/image-rs/image-png/issues/116)). I think the API for dealing with the text chunks can be improved upon, so I welcome any comments on how to make it more "Rustic". In particular, some methods might be performing unnecessary allocs, where a clever use of smart pointers might suffice. I have added documentation, and examples on how to handle text chunks; however, I don't have any tests, and would welcome recommendations on what sort of stuff to test.

nokome commented:
Hi. First thanks to @HeroicKatora and the other contributors for `image-png` 🙏🏽 . Also, thanks to @sayantangkhan for this PR, it is the primary reason we chose `image-png` for our use case - injecting and extracting metadata into PNGs. I'd love to see this PR merged and was wondering if there is anything I can do to help that along?

fintelia commented on src/text_metadata.rs:
```diff
@@ -0,0 +1,546 @@
+
+impl TEXtChunk {
+    /// Constructs a new TEXtChunk.
+    /// Not sure whether it should take &str or String.
+    pub fn new(keyword: &str, text: &str) -> Self {
```
Idiomatically, it would be better for this function to take `String`s. That avoids an extra copy if the caller doesn't need the string afterwards

fintelia commented on src/text_metadata.rs:
```diff
@@ -0,0 +1,546 @@
+
+impl ZTXtChunk {
+    /// Creates a new ZTXt chunk.
+    pub fn new(keyword: &str, text: &str) -> Self {
```
This should also take `String`s

fintelia commented on src/text_metadata.rs:
```diff
@@ -0,0 +1,546 @@
+
+impl ITXtChunk {
+    /// Constructs a new iTXt chunk. Leaves all but keyword and text to default values.
+    pub fn new(keyword: &str, text: &str) -> Self {
```
See above

sayantangkhan replied:
See commit c755f93.

fintelia reviewed (approved):
Left a few comments. If you could also add an entry to CHANGELOG.md (in a new 'Unreleased' section) that would make it easier next time we go to make a release.

fintelia commented:
Thanks for the ping! I somehow lost track of this PR

fintelia commented:
Have you manually verified that text chunks written using this patch can be read by other software, and vice versa?

sayantangkhan commented:
@fintelia Yes, the tEXt chunks I checked using imagemagick, and the zTXt and the iTXt chunks I checked using C++ libraries that have text chunk support.

nokome commented:
I have also manually verified `iTXt` compatibility using `pngchunks` and `pnginfo` from the `pngtools` suite. The `temp.png` file was generated using `encoder.add_itxt_chunk("json", &json)`.

`pngchunks` shows that an `iTXt` is present:

```sh
> pngchunks temp.png
Chunk: Data Length 13 (max 2147483647), Type 1380206665 [IHDR]
  Critical, public, PNG 1.2 compliant, unsafe to copy
  IHDR Width: 800
  IHDR Height: 20
  IHDR Bitdepth: 8
  IHDR Colortype: 6
  IHDR Compression: 0
  IHDR Filter: 0
  IHDR Interlace: 0
  IHDR Compression algorithm is Deflate
  IHDR Filter method is type zero (None, Sub, Up, Average, Paeth)
  IHDR Interlacing is disabled
  Chunk CRC: 1970493774
Chunk: Data Length 147 (max 2147483647), Type 1951945833 [iTXt]
  Ancillary, public, PNG 1.2 compliant, safe to copy
  ... Unknown chunk type
  Chunk CRC: 273375566
Chunk: Data Length 1613 (max 2147483647), Type 1413563465 [IDAT]
  Critical, public, PNG 1.2 compliant, unsafe to copy
  IDAT contains image data
  Chunk CRC: 2033459806
Chunk: Data Length 0 (max 2147483647), Type 1145980233 [IEND]
  Critical, public, PNG 1.2 compliant, unsafe to copy
  IEND contains no data
```

`pnginfo` shows the correct keyword and content of the chunk, but appears to mislabel it as a "xTXt deflate compressed" chunk:

```sh
> pnginfo temp.png
temp.png...
  Image Width: 800 Image Length: 20
  Bitdepth (Bits/Sample): 8
  Channels (Samples/Pixel): 4
  Pixel depth (Pixel Depth): 32
  Colour Type (Photometric Interpretation): RGB with alpha channel 
  Image filter: Single row per byte filter 
  Interlacing: No interlacing 
  Compression Scheme: Deflate method 8, 32k window
  Resolution: 0, 0 (unit unknown)
  FillOrder: msb-to-lsb
  Byte Order: Network (Big Endian)
  Number of text strings: 1
    json (xTXt deflate compressed): {"type":"CodeChunk","programmingLanguage":"python","text":"print('Hello world!')","id":"_KFajWiJDPtCh7rJ8AWhd","outputs":["Hello world!"]}
```

```

HeroicKatora reviewed (commented):
Oh my, thanks for the ping because there has indeed been a misunderstanding on my part. I had previous, a good while ago, started doing an extensive review as below but forgotten to _submit_ that review. Yeah, so it was just visible for me... I'm sorry for failing to give timely feedback here, I should have realized.

HeroicKatora commented on src/decoder/stream.rs:
```diff
@@ -320,14 +349,22 @@ impl From<DecodingError> for io::Error {
     inflater: ZlibStream,
     /// The complete image info read from all prior chunks.
-    pub(crate) info: Option<Info<'static>>,
+    pub info: Option<Info<'static>>,
```
That said, I wouldn't make accepting the PR dependent on this portion.

sayantangkhan commented on src/text_metadata.rs:
```diff
@@ -0,0 +1,546 @@
+    pub optionally_compressed_text: OptCompressed,
+}
+
+/// Enum encoding the compressed and uncompressed states of zTXt/iTXt text field.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum OptCompressed {
+    /// Compressed version of text field. Can be at most 2GB.
+    Compressed(Vec<u8>),
+    /// Uncompressed text field.
+    Uncompressed(String),
+}
```
That was not my original intention: the original intention was just to avoid unnecessary expensive compression/decompression when reading the text multiple times. Another option is to keep the internal text data compressed all the time, decompressing on-demand.

HeroicKatora replied:
It might suffice to have an opaque wrapper around this but in this manner all the variants can be constructed.

sayantangkhan replied:
I made the field private in the commit fb8bf59. The only way the user can set a text field is by creating a new text chunk, and they can get the text from a chunk using the get_text method.

sayantangkhan commented on src/text_metadata.rs:
```diff
@@ -0,0 +1,546 @@
+    pub optionally_compressed_text: OptCompressed,
+}
+
+impl Default for ITXtChunk {
+    fn default() -> Self {
+        Self {
+            keyword: "Default keyword".to_string(),
+            compressed: false,
+            language_tag: String::default(),
+            translated_keyword: String::default(),
+            optionally_compressed_text: OptCompressed::Uncompressed(String::default()),
+        }
+    }
+}
```
This was unnecessary, and I will remove the impl.

sayantangkhan commented on src/text_metadata.rs:
```diff
@@ -0,0 +1,546 @@
+    }
+
+    /// Decompresses the inner text, mutating its own state.
+    pub fn decompress_text(&mut self) -> Result<(), DecodingError> {
```
What would be a reasonable default limit: or does the caller need to provide the limit?

HeroicKatora replied:
Ideally, there would be a separate method where one can also provide a maximum number of bytes to allocate, returning an error when that maximum would be exceeded. It's fine when the simpler method substitutes a fairly large but reasonable limit (e.g. a few mega bytes). Callers that want to handle very large comments can then call the specialized method with `usize::MAX` to accept the risk of OOM.

sayantangkhan commented on src/text_metadata.rs:
```diff
@@ -0,0 +1,546 @@
+    pub fn decompress_text(&mut self) -> Result<(), DecodingError> {
+        match &self.optionally_compressed_text {
+            OptCompressed::Compressed(v) => {
+                let uncompressed_raw = decompress_to_vec_zlib(&v[..])
```
Created a new method in fb8bf59 called `decompress_text_with_limit`, that takes a limiting number of bytes. `decompress_text` calls `decompress_text_with_limit` with a limit of 2MiB.

sayantangkhan commented:
Pinging @HeroicKatora to let them know I've responded to all their review requests except the one about `pub (crate) info` vs `pub info`, for which I don't really have a clean solution.

HeroicKatora reviewed (commented):
Thanks for the ping again. With regards to the open comment, would it fulfill all the needs to have an accessor function instead of `pub info`:

```rust
impl StreamingDecoder {
    pub fn info(&self) -> Option<&Info<'static>> {
        self.info.as_ref()
    }
}
```

I'm somewhat reluctant to make the field open for _writing_ as there are likely some inner invariants that become very much harder to test under that model. My biggest concern is that this makes the interface less reliable and could add more situations where it panics. Even if that requires manually overwriting the field there is nothing inherently discouraging the user from doing so once the field is `pub`.

sayantangkhan commented:
Thanks @HeroicKatora for the suggestion: that indeed did work, and I've switched the field from `pub` to `pub(crate)` in the latest commit, and updated the code examples to use the accessor function.
